dash/include/dash/halo/Halo.h

```diff
@@ -0,0 +1,1062 @@
+#ifndef DASH__HALO__HALO_H__
+#define DASH__HALO__HALO_H__
+
+#include <dash/memory/GlobStaticMem.h>
+#include <dash/iterator/GlobIter.h>
+
+#include <dash/internal/Logging.h>
+
+#include <functional>
+
+namespace dash {
+
+//TODO move to util/FunctionalExpr.h
+template<typename BaseT, typename ExpT>
+constexpr BaseT pow(BaseT base, ExpT exp) {
+  static_assert(std::is_integral<BaseT>::value, "Base must be an integer.");
+  static_assert(std::is_integral<ExpT>::value && std::is_unsigned<ExpT>::value,
+      "Exponent must be an unsigned integer.");
+
+  return (exp == 0 ? 1 : base * pow(base, exp - 1));
+}
+
+
+
+template <dim_t NumDimensions>
+class Stencil : public Dimensional<int16_t, NumDimensions> {
+
+private:
+  using BaseT = Dimensional<int16_t, NumDimensions>;
+
+public:
+  // TODO constexpr
+  Stencil() {
+    for (dim_t i(0); i < NumDimensions; ++i) {
+      this->_values[i] = 0;
+    }
+  }
+
+  template <typename... Values>
+  constexpr Stencil(int16_t value, Values... values) : BaseT::Dimensional(value, values...) {
+  }
+
+  // TODO as constexpr
+  int max() const {
+    int16_t max = 0;
+    for (dim_t i(0); i < NumDimensions; ++i)
+      max = std::max((int) max, (int) std::abs(this->_values[i]));
+    return max;
+  }
+}; // Stencil
+
+
+
+template <dim_t NumDimensions, std::size_t NumStencilPoints>
+class StencilSpec {
+
+private:
+  using SelfT    = StencilSpec<NumDimensions, NumStencilPoints>;
+  using StencilT = Stencil<NumDimensions>;
+  using SpecsT   = std::array<StencilT, NumStencilPoints>;
+
+public:
+  constexpr StencilSpec(const SpecsT& specs) : _specs(specs) {}
+
+  // TODO constexpr
+  StencilSpec(const SelfT& other) { _specs = other._specs; }
+
+  constexpr const SpecsT& stencilSpecs() const { return _specs; }
+
+  static constexpr std::size_t numStencilPoints() { return NumStencilPoints; }
+
+  constexpr const StencilT& operator[](std::size_t index) const { return _specs[index]; }
+
+private:
+  SpecsT _specs{};
+}; // StencilSpec
+
+
+
+/**
+ * Specifies the values global boundary Halos
+ */
+enum class Cycle : uint8_t {
+  /// No global boundary Halos
+  NONE,
```

Comment:
Are we sure there is no name conflict on `dash::NONE`? Can we move this to `dash::Stencil::NONE`?

Reply:
it's an enum class -> you can call it with dash::Cycle::NONE only

Reply:
Ahh, didn't see that (and wasn't aware of enum classes to be honest ^^)